go/pkg/sysdb/coordinator/table_catalog.go

```diff
@@ -859,3 +940,149 @@ func (tc *Catalog) FlushCollectionCompaction(ctx context.Context, flushCollectio
 	}
 	return flushCollectionInfo, nil
 }
+
+// Pre-Context for understanding this method:
+//  1. Information about collection version history is maintained in the VersionFile in S3.
+//  2. The VersionFileName is maintained in the Postgres table.
+//  3. When updating CollectionEntry, a CAS operation against both version and version file name is performed.
+//  4. Since Segment information is maintained in a separate table, a Transaction
+//     is used to atomically update the CollectionEntry and Segment data.
+//
+// Algorithm:
+// 1. Get the collection entry from the table.
+// 2. Prepare the new version file.
+// 3. Write the version file to S3.
+// 4. Till the CAS operation succeeds, retry the operation (i.e. goto 1)
+// 5. 		If version CAS fails - then fail the operation to the Compactor.
+// 6. 		If version file name CAS fails - read updated file and write a new version file to S3.
+func (tc *Catalog) FlushCollectionCompactionForVersionedCollection(ctx context.Context, flushCollectionCompaction *model.FlushCollectionCompaction) (*model.FlushCollectionInfo, error) {
+	// The result that is sent back to the Compactor.
+	flushCollectionInfo := &model.FlushCollectionInfo{
+		ID: flushCollectionCompaction.ID.String(),
+	}
+
+	// Do the operation in a loop until the CollectionEntry is updated,
+	// 		OR FAIL the operation if the version is stale
+	//      OR other DB error.
+	//
+	// In common case, the loop will run only once.
+	// The loop with run more than once only when GC competes to update the
+	// VersionFileName. More precisely, when GC updates the VersionFile in S3
+	// to mark certain versions and then tries to update the VersionFileName in
+	// the table at the same time.
+	for {
+		// Get the current version info and the version file from the table.
+		collectionEntry, err := tc.metaDomain.CollectionDb(ctx).GetCollectionEntry(types.FromUniqueID(flushCollectionCompaction.ID), nil)
+		if err != nil {
+			return nil, err
+		}
+		if collectionEntry == nil {
+			return nil, common.ErrCollectionNotFound
+		}
+		if collectionEntry.IsDeleted {
+			return nil, common.ErrCollectionSoftDeleted
+		}
+
+		versionAtCompactionStart := int64(flushCollectionCompaction.CurrentCollectionVersion)
+		existingVersion := int64(collectionEntry.Version)
+
+		// Do a check to see if the version is stale.
+		if existingVersion > versionAtCompactionStart {
```

Comment:
what happens if `existingVersion < versionAtCompactionStart`. That's likely a system invariant violation?